### PR TITLE
8340409: [lworld] Simple serialization and deserialization of core migrated classes

### PR DESCRIPTION
**Pull request by RogerRiggs**
Serialization and deerialization of value classes. 
For java.base value classes migrated from identity classes:
  - Must be annotated with jdk.internal.MigratedValueClass (in tests too)
  - Must have a constructor or static factory method:
    - annotated with jdk.internal.value.DeserializeConstrucctor
    - constructor parameters must match order and types of serializable fields of the class
  - Value class must not have superclass with fields; only Object or abstract w/o fields
  - Uses same technique as for Records to extract and concat values from stream to invoke constructor
  - Migrated value classes are serialized using the same format/fields as the identity class

Non-migrated value classes throw InvalidClassException

Along the way, refactored code to break out different modes of deserialization: full custom deserialization with custom data, default deserialization only, externalizable, no local class, record, and then added mode to deserialize a value class.

Updates to value class tests and added a combo test to check many combinations of value and identity classes with different parameters.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8340409](https://bugs.openjdk.org/browse/JDK-8340409): [lworld] Simple serialization and deserialization of core migrated classes (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1248/head:pull/1248` \
`$ git checkout pull/1248`

Update a local copy of the PR: \
`$ git checkout pull/1248` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1248/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1248`

View PR using the GUI difftool: \
`$ git pr show -t 1248`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1248.diff">https://git.openjdk.org/valhalla/pull/1248.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1248#issuecomment-2364524312)

### CODE REVIEW

File: src/java.base/share/classes/java/io/ObjectStreamClass.java [hunk 1]
```diff
@@ -1734,9 +1890,10 @@ private static ObjectStreamField[] getDefaultSerialFields(Class<?> cl) {
         ArrayList<ObjectStreamField> list = new ArrayList<>();
         int mask = Modifier.STATIC | Modifier.TRANSIENT;
 
+        int argIndex = 0;
         for (int i = 0; i < clFields.length; i++) {
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/io/ObjectStreamClass.java:1894]:
```suggestion
        for (int i = 0, argIndex = 0; i < clFields.length; i++) {
```

File: src/java.base/share/classes/java/io/ObjectStreamClass.java [hunk 2]
```diff
@@ -1464,6 +1561,67 @@ private ObjectStreamClass getVariantFor(Class<?> cl)
...
+                    try {
+                        m.setAccessible(true);
+                        return (m instanceof Constructor<?> cons)
+                                ? lookup.unreflectConstructor(cons)
+                                : lookup.unreflect(((Method) m));
+                    } catch (IllegalAccessException iae) {
+                        iae.printStackTrace();
+                        return null;
+                    }})
+                .filter(m -> matchFactoryParamTypes(clazz, m, fields))
```
[the remark this answers is not in the document]
Comment (viktorklang-ora) [reply src/java.base/share/classes/java/io/ObjectStreamClass.java:1590]: Yeah, sounds good. 👍

**Comment by RogerRiggs** [comment]
Are there any more review comments; I'd like to get this integrated.

**Pull request closed without merging**